src/DotNetWorker.Core/Converters/Converter/DefaultInputConverterProvider.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.Functions.Worker.Converters
+{
+    /// <summary>
+    /// An implementation of <see cref="IInputConverterProvider"/> to get IInputConverter instances.
+    ///  - Provides IInputConverter instances from what is defined in WorkerOptions.InputConverters
+    ///  - Provides IInputConverter instances when requested for a specific type explicitly.
+    ///  - Internally caches the instances created.
+    /// </summary>
+    internal sealed class DefaultInputConverterProvider : IInputConverterProvider
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly WorkerOptions _workerOptions;       
+         
+        /// <summary>
+        /// Stores all input converters.
+        /// </summary>
+        private readonly ConcurrentDictionary<Type, IInputConverter> _converterCache = new();
+
+        /// <summary>
+        /// Stores the default converter instances.
+        /// This is an ordered sub set of what is present in _converterCache.
+        /// </summary>
+        private IReadOnlyList<IInputConverter> _defaultConverters;
+
+        public DefaultInputConverterProvider(IOptions<WorkerOptions> workerOptions, IServiceProvider serviceProvider)
+        {
+            _workerOptions = workerOptions?.Value ?? throw new ArgumentNullException(nameof(workerOptions));
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            
+            InitializeConverterCacheWithDefaultConverters();
+        }
+                
+        /// <summary>
+        /// Gets an ordered collection of default converter instances.
+        /// </summary>
+        public IEnumerable<IInputConverter> DefaultConverters => _defaultConverters;
+                
+        /// <summary>
+        /// Gets an instance of the converter for the type requested.
+        /// </summary>
+        /// <param name="converterType">The type of IConverter implementation to return.</param>
+        /// <returns>IConverter instance of the requested type.</returns>
+        public IInputConverter GetOrCreateConverterInstance(Type converterType)
+        {
+            if (converterType == null)
+            {
+                throw new ArgumentNullException((nameof(converterType)));
+            }
```

Comment:
Should we check that this implements `IInputConverter`?

Reply:
Good call. Added that check now.